File: Swift/BOJ/10000/14888.swift
//
//  14888.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/03/18.
//

import Foundation

func backtracking(_ current: [Int], _ ops: [Int], _ total: Int) {
    if current.count == total {
        opSet.insert(current)
        return
    }
    
    var ops = ops
    for i in 0..<ops.count {
        if ops[i] > 0 {
            ops[i] -= 1
            backtracking(current + [i], ops, total)
            ops[i] += 1
        }
    }
}

let count = Int(readLine()!)!
let nums = readLine()!.components(separatedBy: " ").map { Int($0)! }
//더하기, 빼기, 곱하기, 나누기
let ops = readLine()!.components(separatedBy: " ").map { Int($0)! }
let total = ops.reduce(0, +)

var opSet: Set<[Int]> = []
backtracking([], ops, total)

var minResult = Int.max, maxResult = Int.min
for ops in opSet {
    var cur = 0
    var number = nums[cur]
    
    for op in ops {
        cur += 1
        
        switch op {
        case 0:
            number += nums[cur]
        case 1:
            number -= nums[cur]
        case 2:
            number *= nums[cur]
        case 3:
            number /= nums[cur]
        default: break
        }
    }

    maxResult = max(maxResult, number)
    minResult = min(minResult, number)
}

print(maxResult)
print(minResult)

File: Python/BOJ/시뮬레이션/14888.py
//
//  14888.py
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/06/05.
//

import sys; readLine = sys.stdin.readline

def backtracking(current, ops):
  if len(current) == n - 1:
    opSet.add(tuple(current))
    return
  
  for i in range(4):
    if ops[i] > 0:
      ops[i] -= 1
      backtracking(current + [i], ops)
      ops[i] += 1

n = int(readLine())
nums = list(map(int, readLine().split()))
ops = list(map(int, readLine().split()))

opSet = set()
backtracking([], ops)

maxResult = -1e9
minResult = 1e9
for ops in opSet:
  number = nums[0]
  for i in range(n-1):
    op = ops[i]
    if op == 0:
      number += nums[i+1]
    elif op == 1:
      number -= nums[i+1]
    elif op == 2:
      number *= nums[i+1]
    elif op == 3:
      number = int(number / nums[i+1])

  maxResult = max(maxResult, number)
  minResult = min(minResult, number)

print(maxResult)
print(minResult)
